src/Symfony/Bundle/FrameworkBundle/Resources/config/messenger.php

```diff
@@ -0,0 +1,179 @@
+use Symfony\Component\Messenger\RoutableMessageBus;
+use Symfony\Component\Messenger\Transport\InMemoryTransportFactory;
+use Symfony\Component\Messenger\Transport\Sender\SendersLocator;
+use Symfony\Component\Messenger\Transport\Serialization\PhpSerializer;
+use Symfony\Component\Messenger\Transport\Serialization\Serializer;
+use Symfony\Component\Messenger\Transport\Serialization\SerializerInterface;
+use Symfony\Component\Messenger\Transport\Sync\SyncTransportFactory;
+use Symfony\Component\Messenger\Transport\TransportFactory;
+
+return static function (ContainerConfigurator $container) {
+    $container->services()
+        // Asynchronous
+        ->set('messenger.senders_locator', SendersLocator::class)
+            ->args([
+                [], // Per message senders map,
+                abstract_arg('senders service locator'), // senders service locator
+            ])
+        ->set('messenger.middleware.send_message', SendMessageMiddleware::class)
+            ->args([
+                service('messenger.senders_locator'),
+                service('event_dispatcher'),
+            ])
+            ->call('setLogger', [service('logger')->ignoreOnInvalid()])
+            ->tag('monolog.logger', ['channel' => 'messenger'])
+
+        // Message encoding/decoding
+        ->set('messenger.transport.symfony_serializer', Serializer::class)
+            ->args([
+                service('serializer'),
+                abstract_arg('format'),
+                [], // Context
+            ])
+        ->alias(SerializerInterface::class, 'messenger.default_serializer')
+
+        ->set('messenger.transport.native_php_serializer', PhpSerializer::class)
+
+        // Middleware
+        ->set('messenger.middleware.handle_message', HandleMessageMiddleware::class)
+            ->abstract()
+            ->args([
+                abstract_arg('bus handler resolver'),
+            ])
+            ->tag('monolog.logger', ['channel' => 'messenger'])
+            ->call('setLogger', [service('logger')->ignoreOnInvalid()])
+
+        ->set('messenger.middleware.add_bus_name_stamp_middleware', AddBusNameStampMiddleware::class)
+            ->abstract()
+
+        ->set('messenger.middleware.dispatch_after_current_bus', DispatchAfterCurrentBusMiddleware::class)
+
+        ->set('messenger.middleware.validation', ValidationMiddleware::class)
+            ->args([
+                service('validator'),
+            ])
+
+        ->set('messenger.middleware.reject_redelivered_message_middleware', RejectRedeliveredMessageMiddleware::class)
+
+        ->set('messenger.middleware.failed_message_processing_middleware', FailedMessageProcessingMiddleware::class)
+
+        ->set('messenger.middleware.traceable', TraceableMiddleware::class)
+            ->abstract()
+            ->args([
+                service('debug.stopwatch'),
+            ])
+
+        // Discovery
+        ->set('messenger.receiver_locator')
+            ->args([
+                [],
```

Comment:
abstract_arg pls

Reply:
you can put "receiver name to service mapping"

Reply:
test fail with this changement

```
Invalid definition for service "messenger.receiver_locator": an array of references is expected as first argument when the "container.service_locator" tag is set.
```

Reply:
The problem is probably that the compiler pass to handle container.service_locator tags, validates the argument. But the service is not used at all and would be removed. So the validation should skip abstract_arg. I wonder if the compiler pass does not run after unused services are removed?

Anyway, please don't use abstract_arg then for messenger.receiver_locator and messenger.retry_strategy_locator. Fixing this should be separate thing.